File: services/knowledge-graph/src/knowledge_graph_mcp/tools/learning.py
"""propose_knowledge tool — submit discovered patterns as candidates."""

import json
from typing import Any, Callable, Awaitable
from knowledge_graph_mcp.db import KnowledgeDB


async def _propose_knowledge_impl(
    category: str,
    key: str,
    content: str,
    scope: str,
    rationale: str,
    db: KnowledgeDB,
    embed_fn: Callable[[str, str, str], Awaitable[list[float]]],
    project_id: str,
    location: str,
    tags: list[str] | None = None,
    session_id: int | None = None,
) -> dict[str, Any]:
    """Submit a knowledge candidate for review."""
    embedding = await embed_fn(content, project_id, location)

    rows = await db.query(
        """INSERT INTO knowledge_candidates
           (category, key, content, embedding, scope, tags, rationale, session_id)
           VALUES ($1, $2, $3, $4::vector, $5, $6::text[], $7, $8)
           RETURNING id""",
        [category, key, content, str(embedding), scope, tags or [], rationale, session_id],
    )
    return {"candidate_id": rows[0]["id"], "status": "pending"}


def register_learning_tools(mcp, get_db, get_settings, embed_fn):
    @mcp.tool()
    async def propose_knowledge(
        category: str,
        key: str,
        content: str,
        scope: str,
        rationale: str,
        tags: list[str] | None = None,
        session_id: int | None = None,
    ) -> str:
        """Submit a discovered pattern/rule as a candidate for review. Include rationale for why it should become permanent knowledge."""
        s = get_settings()
        result = await _propose_knowledge_impl(
            category, key, content, scope, rationale,
            get_db(), embed_fn, s.gcp_project_id, s.gcp_location,
            tags=tags, session_id=session_id,
        )
        return json.dumps(result, default=str)

File: services/knowledge-graph/src/tools/learning.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { eq, and, desc } from "drizzle-orm";
import type { Db } from "../db/index.js";
import type { Config } from "../config.js";
import { embedText } from "../embeddings.js";
import { knowledgeCandidates } from "../db/schema.js";

export function registerLearningTools(
  server: McpServer,
  db: Db,
  config: Config,
) {
  server.tool(
    "propose_knowledge",
    "Propose a new knowledge entry for curator review",
    {
      category: z.string().describe("Knowledge category"),
      key: z.string().describe("Unique key for this entry"),
      content: z.string().describe("Knowledge content"),
      scope: z.string().describe("Scope identifier"),
      rationale: z
        .string()
        .describe("Why this should become permanent knowledge"),
      tags: z.array(z.string()).optional().describe("Topic tags"),
      session_id: z.number().optional().describe("Associated session ID"),
    },
    async ({ category, key, content, scope, rationale, tags, session_id }) => {
      const embedding = await embedText(
        content,
        config.gcpProjectId,
        config.gcpLocation,
      );

      const [candidate] = await db
        .insert(knowledgeCandidates)
        .values({
          category,
          key,
          content,
          embedding,
          rationale,
          scope,
          tags: tags ?? [],
          sessionId: session_id,
        })
        .returning({ id: knowledgeCandidates.id });

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({
              candidate_id: candidate.id,
              status: "pending",
            }),
          },
        ],
      };
    },
  );

  server.tool(
    "list_candidates",
    "List knowledge candidates with optional status filter",
    {
      status: z
        .enum(["pending", "approved", "rejected"])
        .optional()
        .default("pending")
        .describe("Filter by status (default: pending)"),
      scope: z.string().optional().describe("Filter by scope"),
      limit: z.number().optional().default(50).describe("Max results"),
    },
    async ({ status, scope, limit }) => {
      const conditions: ReturnType<typeof eq>[] = [];
      if (status) conditions.push(eq(knowledgeCandidates.status, status));
      if (scope) conditions.push(eq(knowledgeCandidates.scope, scope));

      const results = await db
        .select({
          id: knowledgeCandidates.id,
          category: knowledgeCandidates.category,
          key: knowledgeCandidates.key,
          content: knowledgeCandidates.content,
          rationale: knowledgeCandidates.rationale,
          scope: knowledgeCandidates.scope,
          tags: knowledgeCandidates.tags,
          status: knowledgeCandidates.status,
          createdAt: knowledgeCandidates.createdAt,
        })
        .from(knowledgeCandidates)
        .where(conditions.length > 0 ? and(...conditions) : undefined)
        .orderBy(desc(knowledgeCandidates.createdAt))
        .limit(limit ?? 50);

      return {
        content: [
          { type: "text" as const, text: JSON.stringify(results, null, 2) },
        ],
      };
    },
  );
}
